src/qibo/models/encodings.py

```diff
@@ -1787,3 +1973,216 @@ def _add_wbd_gate(
+
+    return qgates
+
+
+def _perm_pair_flip_ops(n: int, m: int, backend=None):
+    """Implement σ_{i,2} as X fan‑in + MCX + X fan‑out."""
+    backend = _check_backend(backend)
+    # let us flip the first qubit when the last {int(n-math.log2(2*m))} qubits are all in the state |0⟩
+    prefix = int(backend.np.ceil(backend.np.log2(2 * m)))
+    x_qubits = list(range(prefix, n))
+    qgates = [gates.X(n - q - 1) for q in x_qubits]
+    qgates.append(
+        gates.X(n - 1).controlled_by(*range(n - int(np.ceil(math.log2(2 * m)))))
+    )  # flip qubit 0
+    qgates.extend(gates.X(n - q - 1) for q in x_qubits)
+
+    return qgates
+
+
+def permutation_synthesis(
+    sigma: list[int],
+    m: int = 2,
+    backend=None,
+):
+    """Given permutation sigma on {0,...,d-1} and a power‑of‑two budget m,
+    factor sigma into the fewest layers sigma_1, sigma_2, ..., sigma_t such that
+        - each layer has at most m disjoint transpositions
+        - each layer moves a power‑of‑two number of indices.
+        Return a circuit synthesis of sigma.
+
+    Args:
+        sigma (list[int] or tuple[int]): permutation description on {0,...,d-1}.
+        m (int): power‑of‑two budget m (Default:2)
+
+    Returns:
+        :class:`qibo.models.circuit.Circuit`: Circuit that implements the permutation sigma.
+
+    References:
+        1. L. Li, J. Luo,
+        *Nearly Optimal Circuit Size for Sparse Quantum State Preparation*
+        `arXiv:2406.16142 (2024) <https://doi.org/10.48550/arXiv.2406.16142>`_.
+    """
+    backend = _check_backend(backend)
+
+    if isinstance(sigma, tuple):
+        sigma = list(sigma)
+
+    if not isinstance(sigma, list):
+        raise_error(TypeError, f"Permutation ``sigma`` must be a ``list`` of ``int``.")
+
+    n = int(backend.np.ceil(backend.np.log2(len(sigma))))
+    if sum([abs(s - i) for s, i in zip(sorted(sigma), range(2**n))]) != 0:
+        raise_error(TypeError, "Permutation sigma must contain all indices {0,...,n-1}")
+
+    if m > 0 and (m & (m - 1)) != 0:
+        raise_error(TypeError, f"budget m must be a power‑of‑two")
+
+    from qibo.quantum_info.utils import (  # pylint: disable=import-outside-toplevel
+        decompose_permutation,
+    )
+
+    # factor sigma into the fewest layers such that
+    # each layer has at most m disjoint transpositions, and
+    # each layer moves a power‑of‑two number of indices
+    layers = decompose_permutation(sigma, m)
+
+    circ_gates = []
+    # in case we have more than one permutation to do, do it in layers
+    for layer in layers:
+        m = len(layer)
+        ell, col_gates, A = _perm_column_ops(layer, n, backend)
+        row_gates = _perm_row_ops(A, ell, m, n, backend)
+        flip_gates = _perm_pair_flip_ops(n, m, backend)
+
+        circ_gates += (
+            col_gates + row_gates + flip_gates + list(reversed(col_gates + row_gates))
+        )
+
+    circ = Circuit(n)
+    circ.add(circ_gates)
+    return circ
```

Comment:
```suggestion
    circuit = Circuit(n)
    # in case we have more than one permutation to do, do it in layers
    for layer in layers:
        m = len(layer)
        ell, col_gates, A = _perm_column_ops(layer, n, backend)
        row_gates = _perm_row_ops(A, ell, m, n, backend)
        flip_gates = _perm_pair_flip_ops(n, m, backend)

        circuit.add(col_gates)
        circuit.add(row_gates)
        circuit.add(flip_gates)
        circuit.add((col_gates + row_gates)[::-1])

    return circuit
```